lib/babe/verify.go

```diff
@@ -327,62 +327,58 @@ func (b *verifier) verifyAuthorshipRight(header *types.Header) error {
 		return ErrBadSignature
 	}
 
-	// check if the producer has equivocated, ie. have they produced a conflicting block?
-	// hashes is hashes of all blocks with same block number as header.Number
-	hashes := b.blockState.GetAllBlocksAtDepth(header.ParentHash)
+	equivocated, err := b.verifyBlockEquivocation(header)
+	if err != nil {
+		return fmt.Errorf("could not verify block equivocation for header %s: %w", header.Hash(), err)
+	}
+	if equivocated {
+		return ErrProducerEquivocated
+	}
 
-	for _, currentHash := range hashes {
-		currentHeader, err := b.blockState.GetHeader(currentHash)
-		if err != nil {
-			return fmt.Errorf("failed get header %s", err)
-		}
+	return nil
+}
 
-		currentBlockProducerIndex, err := getAuthorityIndex(currentHeader)
-		if err != nil {
-			return fmt.Errorf("failed to get authority index %s", err)
-		}
+// verifyBlockEquivocation checks if given block's author has occupied corresponding slot more than once.
+// It returns true if block was equivocated.
+func (b *verifier) verifyBlockEquivocation(header *types.Header) (bool, error) {
```

Comment:
Can you add a unit test for this new method?

Reply:
`Test_verifier_verifyAuthorshipRightEquivocatory` and `TestVerifyAuthorshipRight_Equivocation`

Reply:
- `Test_verifier_verifyAuthorshipRightEquivocatory` tests its caller `verifyAuthorshipRightEquivocatory` not `verifyBlockEquivocation`. It does not cover all code paths in that function either.
- `TestVerifyAuthorshipRight_Equivocation` takes 4 seconds to execute and is an integration test, not a unit test. And also does not cover all code paths in `verifyBlockEquivocation`

Reply:
added the test